### PR TITLE
bpo-1635741: In pickle module, inject module state from class methods

### PR DESCRIPTION
**Pull request by koubaa**
This PR prepares for changing to heap types.
When switching to heap types, the clinic can be used to get the module state.

https://bugs.python.org/issue1635741

### CODE REVIEW

**Comment by koubaa** [comment]
@vstinner @corona10 @shihai1991 please review

File: Modules/_pickle.c
Comment (koubaa) [inline Modules/_pickle.c]: PickleState is not used here, but is added as an argument to avoid complicating the macro/switch case in 
```
static PyObject *
load(UnpicklerObject *self, PickleState *st)
```

**Comment by tiran** [comment]
Most modules are using ``state`` for the state variable. Could you please replace ``st`` with ``state`` so we have a consistent naming scheme across the code base?

**Comment by koubaa** [comment]
> Most modules are using `state` for the state variable. Could you please replace `st` with `state` so we have a consistent naming scheme across the code base?

@tiran I did that in my initial attempt [here](https://github.com/python/cpython/pull/23188).  The reason I didn't do it here is because there are local variables named state in some functions (see for example line 3949 in save_reduce.  Do you have a suggestion on what to change those names to?  I don't understand exactly what those variables represent

File: Modules/_pickle.c
Comment (shihai1991) [inline Modules/_pickle.c]: nitpick: Pls add parentheses in here.

File: Modules/_pickle.c
Comment (shihai1991) [inline Modules/_pickle.c]: ditto

**Review by shihai1991 — COMMENTED** [review]
> The reason I didn't do it here is because there are local variables named state in some functions (see for example line 3949 in save_reduce.

the `state` would be more exact. If there have other arguments have own this name, keep `st`  unchanged is not a big probleam.

File: Modules/_pickle.c
Comment (shihai1991) [reply Modules/_pickle.c]: some other lines should be updated too.

File: Modules/_pickle.c
Comment (vstinner) [inline Modules/_pickle.c]: please rename "S" to "state", and move it at the beginning.

File: Modules/_pickle.c
Comment (vstinner) [inline Modules/_pickle.c]: Would it be possible to pass the state as the first argument, as passing a context? Same remark for the whole PR.

File: Modules/_pickle.c
Comment (vstinner) [reply Modules/_pickle.c]: I guess that you mean braces. Yeah, I agree: please add { ... } (PEP 7).

File: Modules/_pickle.c
Comment (koubaa) [reply Modules/_pickle.c]: @vstinner sure, I didn't want to break the CPython convention of passing self first but if this is your preference I'll change it

**Comment by koubaa** [comment]
@shihai1991 @vstinner I addressed these comments

**Review by shihai1991 — APPROVED** [review]
This PR is huge :) . About adding parenthesis, I will update it when those code have relation with the bpo.

**Comment by iritkatriel** [comment]
https://bugs.python.org/issue1635741 is closed. What is the status of this PR?

**Comment by vstinner** [comment]
The change is still relevant, but should use a new issue number.

Moreover, the SC asked to put the conversion of static types to heap types on hold. @encukou and @erlend-aasland wrote https://peps.python.org/pep-0687/ which may unblock the situation but it's still a draft.

**Comment by erlend-aasland** [comment]
> The change is still relevant, but should use a new issue number.
> 
> Moreover, the SC asked to put the conversion of static types to heap types on hold. @encukou and @erlend-aasland wrote https://peps.python.org/pep-0687/ which may unblock the situation but it's still a draft.

FYI, PEP 687 was accepted.

**Comment by encukou** [comment]
Pickle module state was isolated in #102982. `koubaa` is listed as co-author, so I hope some of this work made it in at the end.

**Comment by erlend-aasland** [comment]
> Pickle module state was isolated in #102982. `koubaa` is listed as co-author, so I hope some of this work made it in at the end.

It did; PR #102982 was based off of Mohamed's work (IIRC, I cherry-picked the commits from this PR onto my new branch).

**Pull request closed without merging**
